File: app/schemas/task.py
"""Pydantic schemas for task and label endpoints."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.project_state import StateGroup
from app.models.task import IssueType, TaskPriority, TaskStatus
from app.models.task_link import LinkType

# --------------------------------------------------------------------------- #
# Label schemas                                                                 #
# --------------------------------------------------------------------------- #


class LabelCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    color: str | None = Field(None, pattern=r"^#[0-9A-Fa-f]{6}$")


class LabelOut(BaseModel):
    id: uuid.UUID
    organization_id: uuid.UUID
    name: str
    color: str | None

    model_config = {"from_attributes": True}


# --------------------------------------------------------------------------- #
# Task request schemas                                                          #
# --------------------------------------------------------------------------- #


class TaskCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    description: str | None = Field(None, max_length=5000)
    state_id: uuid.UUID | None = None
    status: TaskStatus = TaskStatus.TODO  # legacy fallback
    priority: TaskPriority = TaskPriority.MEDIUM
    issue_type: IssueType = IssueType.TASK
    assignee_user_id: uuid.UUID | None = None
    parent_task_id: uuid.UUID | None = None
    sprint_id: uuid.UUID | None = None
    module_id: uuid.UUID | None = None
    position: int = Field(0, ge=0)
    story_points: int | None = Field(None, ge=0, le=100)
    start_date: datetime | None = None
    due_date: datetime | None = None


class TaskUpdateRequest(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=500)
    description: str | None = None
    state_id: uuid.UUID | None = None
    status: TaskStatus | None = None  # legacy fallback
    priority: TaskPriority | None = None
    issue_type: IssueType | None = None
    assignee_user_id: uuid.UUID | None = None
    parent_task_id: uuid.UUID | None = None
    sprint_id: uuid.UUID | None = None
    module_id: uuid.UUID | None = None
    position: int | None = Field(None, ge=0)
    story_points: int | None = Field(None, ge=0, le=100)
    start_date: datetime | None = None
    due_date: datetime | None = None


# --------------------------------------------------------------------------- #
# Task response schema                                                          #
# --------------------------------------------------------------------------- #


class TaskStateOut(BaseModel):
    """Embedded state info in task responses."""
    id: uuid.UUID
    name: str
    color: str
    group: StateGroup

    model_config = {"from_attributes": True}


class TaskSummary(BaseModel):
    """Lightweight task reference for parent/subtask/link contexts."""
    id: uuid.UUID
    title: str
    status: TaskStatus
    issue_type: IssueType
    sequence_id: int | None = None
    state: TaskStateOut | None = None

    model_config = {"from_attributes": True}


class TaskLinkOut(BaseModel):
    id: uuid.UUID
    link_type: LinkType
    source_task: TaskSummary
    target_task: TaskSummary
    created_at: datetime

    model_config = {"from_attributes": True}


class TaskLinkCreateRequest(BaseModel):
    target_task_id: uuid.UUID
    link_type: LinkType


# --------------------------------------------------------------------------- #
# Bulk update schema                                                             #
# --------------------------------------------------------------------------- #


class BulkTaskUpdates(BaseModel):
    state_id: uuid.UUID | None = None
    priority: TaskPriority | None = None
    assignee_user_id: uuid.UUID | None = None
    sprint_id: uuid.UUID | None = None


class BulkUpdateRequest(BaseModel):
    task_ids: list[uuid.UUID] = Field(..., min_length=1, max_length=500)
    updates: BulkTaskUpdates


class BulkUpdateResponse(BaseModel):
    updated_count: int


# --------------------------------------------------------------------------- #
# CSV import response schema                                                     #
# --------------------------------------------------------------------------- #


class CSVImportResponse(BaseModel):
    imported_count: int
    errors: list[str] = []


class TaskOut(BaseModel):
    id: uuid.UUID
    organization_id: uuid.UUID
    project_id: uuid.UUID
    assignee_user_id: uuid.UUID | None
    created_by_user_id: uuid.UUID | None = None
    parent_task_id: uuid.UUID | None
    sprint_id: uuid.UUID | None
    module_id: uuid.UUID | None
    state_id: uuid.UUID | None = None
    sequence_id: int | None = None
    title: str
    description: str | None
    state: TaskStateOut | None = None
    status: TaskStatus
    priority: TaskPriority
    issue_type: IssueType
    position: int
    story_points: int | None
    start_date: datetime | None
    due_date: datetime | None
    labels: list[LabelOut] = []
    parent: TaskSummary | None = None
    subtasks: list[TaskSummary] = []
    links: list[TaskLinkOut] = []
    deleted_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/task.ts
export type TaskStatus = "todo" | "in_progress" | "done" | "blocked"
export type TaskPriority = "low" | "medium" | "high" | "urgent"
export type IssueType = "bug" | "story" | "epic" | "task" | "subtask"

export interface Label {
  id: string
  organization_id: string
  name: string
  color: string | null
}

export type LinkType = "blocks" | "is_blocked_by" | "relates_to" | "duplicate_of"

export interface TaskSummary {
  id: string
  title: string
  status: TaskStatus
  issue_type: IssueType
}

export interface TaskLinkOut {
  id: string
  link_type: LinkType
  source_task: TaskSummary
  target_task: TaskSummary
  created_at: string
}

export interface Task {
  id: string
  organization_id: string
  project_id: string
  assignee_user_id: string | null
  created_by_user_id: string | null
  parent_task_id: string | null
  sprint_id: string | null
  module_id: string | null
  title: string
  description: string | null
  status: TaskStatus
  priority: TaskPriority
  issue_type: IssueType
  position: number
  story_points: number | null
  start_date: string | null
  due_date: string | null
  labels: Label[]
  parent: TaskSummary | null
  subtasks: TaskSummary[]
  links: TaskLinkOut[]
  deleted_at: string | null
  created_at: string
  updated_at: string
}

export interface TaskFilters {
  status?: TaskStatus
  priority?: TaskPriority
  assignee_user_id?: string
  issue_type?: IssueType
  sprint_id?: string
  no_sprint?: boolean
}

export const TASK_STATUS_LABELS: Record<TaskStatus, string> = {
  todo: "Todo",
  in_progress: "In Progress",
  done: "Done",
  blocked: "Blocked",
}

export const TASK_PRIORITY_LABELS: Record<TaskPriority, string> = {
  low: "Low",
  medium: "Medium",
  high: "High",
  urgent: "Urgent",
}

export const ISSUE_TYPE_LABELS: Record<IssueType, string> = {
  bug: "Bug",
  story: "Story",
  epic: "Epic",
  task: "Task",
  subtask: "Sub-task",
}

export const LINK_TYPE_LABELS: Record<LinkType, string> = {
  blocks: "Blocks",
  is_blocked_by: "Is blocked by",
  relates_to: "Relates to",
  duplicate_of: "Duplicate of",
}

export const KANBAN_COLUMNS: TaskStatus[] = ["todo", "in_progress", "done", "blocked"]
